File: VideoStreamingBackend/src/application/use-cases/GetStreamStatus.ts
/**
 * Use case: Get stream status for a camera.
 * Returns session info including WebRTC signaling URL.
 */
import { StreamSession } from '../../domain/entities/StreamSession';
import { StreamManager } from '../../infrastructure/streaming/StreamManager';

export interface StreamStatusResult {
  cameraId: string;
  isStreaming: boolean;
  status: string;
  session: StreamSession | null;
  /** WebRTC signaling URL (replaces old websocket_url for JPEG frames) */
  signalingUrl: string | null;
}

export class GetStreamStatusUseCase {
  constructor(private readonly streamManager: StreamManager) {}

  execute(cameraId: string): StreamStatusResult {
    const session = this.streamManager.getSession(cameraId) ?? null;
    const isStreaming = this.streamManager.isStreaming(cameraId);

    return {
      cameraId,
      isStreaming,
      status: session?.status ?? 'stopped',
      session,
      signalingUrl: isStreaming ? `/ws/signaling/${cameraId}` : null,
    };
  }
}

File: StreamingBackend/internal/application/usecases/get_stream_status.go
package usecases

import (
	"fmt"

	"github.com/vigileye/streaming-backend/internal/domain/entities"
	"github.com/vigileye/streaming-backend/internal/infrastructure/config"
	"github.com/vigileye/streaming-backend/internal/infrastructure/streaming"
)

// StreamStatusResult contains stream status information.
type StreamStatusResult struct {
	CameraID     string
	IsStreaming  bool
	Status       string
	Session      *entities.StreamSession
	SignalingURL string
	WHEPEndpoint string // Direct WHEP URL for MediaMTX viewers
}

// GetStreamStatusUseCase handles getting stream status.
type GetStreamStatusUseCase struct {
	streamManager *streaming.StreamManager
	cfg           *config.Config
}

// NewGetStreamStatusUseCase creates a new GetStreamStatusUseCase.
func NewGetStreamStatusUseCase(sm *streaming.StreamManager, cfg *config.Config) *GetStreamStatusUseCase {
	return &GetStreamStatusUseCase{
		streamManager: sm,
		cfg:           cfg,
	}
}

// Execute returns the stream status for a camera.
func (uc *GetStreamStatusUseCase) Execute(cameraID string) *StreamStatusResult {
	session, _ := uc.streamManager.GetSession(cameraID)
	isStreaming := uc.streamManager.IsStreaming(cameraID)

	status := "stopped"
	if session != nil {
		status = string(session.Status)
	}

	signalingURL := ""
	whepEndpoint := ""
	if isStreaming {
		signalingURL = "/api/v1/webrtc/offer"

		if uc.streamManager.IsMediaMTXEnabled() {
			if uc.streamManager.IsCameraUsingMediaMTX(cameraID) {
				// Provide direct WHEP endpoint for frontend
				whepEndpoint = fmt.Sprintf("%s/%s/whep", uc.cfg.MediaMTXWHEPURL, cameraID)
			}
		}
	}

	return &StreamStatusResult{
		CameraID:     cameraID,
		IsStreaming:  isStreaming,
		Status:       status,
		Session:      session,
		SignalingURL: signalingURL,
		WHEPEndpoint: whepEndpoint,
	}
}
